src/soc/ibm/power9/occ.c

```diff
@@ -0,0 +1,238 @@
+
+static void pm_ocb_setup(uint32_t ocb_bar)
+{
+	write_scom(PU_OCB_PIB_OCBCSR0_OR, PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_MODE));
+	write_scom(PU_OCB_PIB_OCBCSR0_CLEAR, PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_TYPE));
+	write_scom(PU_OCB_PIB_OCBAR0, (uint64_t)ocb_bar << 32);
+}
+
+static void check_ocb_mode(uint64_t OCBCSR_address, uint64_t OCBSHCS_address)
+{
+	uint64_t ocb_pib = read_scom(OCBCSR_address);
+
+	/*
+	 * The following check for circular mode is an additional check
+	 * performed to ensure a valid data access.
+	 */
+	if ((ocb_pib & PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_MODE)) &&
+	    (ocb_pib & PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_TYPE))) {
+		/*
+		 * Check if push queue is enabled. If not, let the store occur
+		 * anyway to let the PIB error response return occur. (That is
+		 * what will happen if this checking code were not here.)
+		 */
+		uint64_t stream_push_control = read_scom(OCBSHCS_address);
+
+		if (stream_push_control & PPC_BIT(OCB_OCI_OCBSHCS0_PUSH_ENABLE)) {
+			uint8_t counter = 0;
+			for (counter = 0; counter < 4; counter++) {
+				/* Proceed if the OCB_OCI_OCBSHCS0_PUSH_FULL is clear */
+				if (!(stream_push_control & PPC_BIT(OCB_OCI_OCBSHCS0_PUSH_FULL)))
+					break;
+
+				stream_push_control = read_scom(OCBSHCS_address);
+			}
+
+			if (counter == 4)
+				die("Failed to write to circular buffer.\n");
+		}
+	}
+}
+
+static void put_ocb_indirect(uint32_t ocb_req_length, uint32_t oci_address,
+			     uint64_t *ocb_buffer)
+{
+	write_scom(PU_OCB_PIB_OCBAR0, (uint64_t)oci_address << 32);
+
+	check_ocb_mode(PU_OCB_PIB_OCBCSR0_RO, PU_OCB_OCI_OCBSHCS0_SCOM);
+
+	for (uint32_t index = 0; index < ocb_req_length; index++)
+		write_scom(PU_OCB_PIB_OCBDR0, ocb_buffer[index]);
+}
+
+static void get_ocb_indirect(uint32_t ocb_req_length, uint32_t oci_address,
+			     uint64_t *ocb_buffer)
+{
+	write_scom(PU_OCB_PIB_OCBAR0, (uint64_t)oci_address << 32);
+	for (uint32_t loopCount = 0; loopCount < ocb_req_length; loopCount++)
+		ocb_buffer[loopCount] = read_scom(PU_OCB_PIB_OCBDR0);
+}
+
+void writeOCCSRAM(uint32_t address, uint64_t * buffer, size_t data_length)
+{
+	pm_ocb_setup(address);
+	put_ocb_indirect(data_length / 8, address, buffer);
+}
+
+void readOCCSRAM(uint32_t address, uint64_t * buffer, size_t data_length)
+{
+	pm_ocb_setup(address);
+	get_ocb_indirect(data_length / 8, address, buffer);
+}
+
+void write_occ_command(uint64_t write_data)
+{
+	check_ocb_mode(PU_OCB_PIB_OCBCSR1_RO, PU_OCB_OCI_OCBSHCS1_SCOM);
+	write_scom(PU_OCB_PIB_OCBDR1, write_data);
+}
+
+void clear_occ_special_wakeups(uint64_t cores)
+{
+	for (size_t i = 0; i < MAX_CORES_PER_CHIP; i += 2) {
+		if (!IS_EX_FUNCTIONAL(i, cores))
+			continue;
+		scom_and_for_chiplet(EC00_CHIPLET_ID + i, EX_PPM_SPWKUP_OCC, ~PPC_BIT(0));
+	}
+}
+
+void special_occ_wakeup_disable(uint64_t cores)
+{
+	enum { PPM_SPWKUP_FSP = 0x200F010B };
+
+	for (int i = 0; i < MAX_CORES_PER_CHIP; ++i) {
+		uint32_t spwkup_addr = PPM_SPWKUP_FSP + 0x01000000 * i;
```

Comment:
This line is useless, `*_for_chiplet()` functions already add proper offset. Remove it and use `PPM_SPWKUP_FSP` instead of `spwkup_addr`